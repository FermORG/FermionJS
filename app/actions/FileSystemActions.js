// @flow
export const SET_ACTIVE_COMPONENT = 'SET_ACTIVE_COMPONENT';

export function setActiveComponent(component: string) {
  return {
    type: SET_ACTIVE_COMPONENT,
    component,
  };
}
