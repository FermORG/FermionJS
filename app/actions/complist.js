//@flow
//complist actions.


// import type { addComponentsStateType } from '../reducers/complist';

// type actionType = {
//   +type: string,
// };

export const ADD_COMPONENTS = 'ADD_COMPONENTS';

export const ADD_TO_PROJECT = 'ADD_TO_PROJECT';

export function AddComponents(components) {
  return {
    type: 'ADD_COMPONENTS',
    components
  };
}

export function AddToProject(component) {
  return {
    type: 'ADD_TO_PROJECT',
    component
  };
}
