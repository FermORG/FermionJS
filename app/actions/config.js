// @flow
// config actions
export const ADD_STATE = 'ADD_STATE';
export const ADD_PROPS = 'ADD_PROPS';
export const ADD_STYLES = 'ADD_STYLES';
export const ADD_EVENTS = 'ADD_EVENTS';
export const DELETE_STATE = 'DELETE_STATE';
export const DELETE_PROPS = 'DELETE_PROPS';
export const DELETE_STYLES = 'DELETE_STYLES';
export const DELETE_EVENTS = 'DELETE_EVENTS';

export function addState(aState, component) {
  return {
    type: 'ADD_STATE',
    aState,
  };
}

export function addProps(prop, component) {
  return {
    type: 'ADD_PROPS',
    prop,
    component
  };
}

export function addStyles(style, component) {
  return {
    type: 'ADD_STYLES',
    style,
    component
  };
}

export function addEvents(event, component) {
  return {
    type: 'ADD_EVENTS',
    event,
    component
  };
}

export function deleteState(stateKey, component) {
  return {
    type: 'DELETE_STATE',
    propKey: stateKey,
  };
}
export function deleteProps(propKey, component) {
  return {
    type: 'DELETE_PROPS',
    prop: propKey,
    component
  };
}
export function deleteStyles(styleKey, component) {
  return {
    type: 'DELETE_STYLES',
    style: styleKey,
    component
  };
}
export function deleteEvents(eventKey, component) {
  return {
    type: 'DELETE_EVENTS',
    event: eventKey,
    component
  };
}
