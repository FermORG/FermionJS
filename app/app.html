<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Fermion</title>
    <script>
      (function() {
        if (!process.env.HOT) {
          const link = document.createElement('link');
          link.rel = 'stylesheet';
          link.href = './dist/style.css';
          // HACK: Writing the script path should be done with webpack
          document.getElementsByTagName('head')[0].appendChild(link);
        }
      }());
    </script>
<link href="https://fonts.googleapis.com/css?family=Concert+One|Hind|Inconsolata|Josefin+Slab|Julius+Sans+One|Lato|Meera+Inimai|Noto+Sans|Open+Sans|Poiret+One|Raleway|Roboto|Saira+Semi+Condensed|Ubuntu|Ubuntu+Condensed|Bungee" rel="stylesheet">
  </head>
  <body>
    <div id="root"></div>
    <script>
      {
        const scripts = [];

        // Dynamically insert the DLL script in development env in the
        // renderer process
        if (process.env.NODE_ENV === 'development') {
          scripts.push('../dll/renderer.dev.dll.js');
        }

        // Dynamically insert the bundled app script in the renderer process
        const port = process.env.PORT || 1212;
        scripts.push(
          (process.env.HOT)
            ? 'http://localhost:' + port + '/dist/renderer.dev.js'
            : './dist/renderer.prod.js'
        );

        document.write(
          scripts
            .map(script => '<script defer src="' + script + '"><\/script>')
            .join('')
        );
      }
    </script>
  </body>
</html>
