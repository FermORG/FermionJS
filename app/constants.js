export default {
  WORKSPACE_ID: 'workspace'
};