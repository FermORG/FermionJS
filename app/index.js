import React from 'react';
import { render } from 'react-dom';
import { AppContainer } from 'react-hot-loader';
import Root from './containers/Root';
import { configureStore, history } from './store/configureStore';
import './app.global.css';
import { ipcRenderer } from 'electron';

ipcRenderer.on('ConfigApp', (event, data) => {
  data = JSON.parse(data);

  const initialState = {
    availableComponents: data.availableComponents,
  }
 const store = configureStore(initialState);

  render(
    <AppContainer>
      <Root store={store} history={history} />
    </AppContainer>,
    document.getElementById('root')
  );

  if (module.hot) {
    module.hot.accept('./containers/Root', () => {
      const NextRoot = require('./containers/Root'); // eslint-disable-line global-require
      render(
        <AppContainer>
          <NextRoot store={store} history={history} />
        </AppContainer>,
        document.getElementById('root')
      );
    });
  }

});
