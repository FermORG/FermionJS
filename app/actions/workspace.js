import { WORKSPACE_ID } from './../constants';

export const ADD_CHILD = 'ADD_CHILD';
export const CREATE_CHILD = 'CREATE_CHILD';
export const DELETE_CHILD = 'DELETE_CHILD';
export const MOVE_CHILD = 'MOVE_CHILD';
export const REMOVE_CHILD = 'REMOVE_CHILD';
export const UPDATE_STYLE = 'UPDATE_STYLE';

const createChild = ({ name, props }) => ({
  name,
  props,
  children: [],
  parentID: WORKSPACE_ID,
});

// export function addChild(targetID, newComponentName) {
//   const newComponent = createChild(newComponentName);
export function addChild(targetID, component) {
  const newComponent = createChild(component);
  return {
    type: ADD_CHILD,
    newComponent,
    targetID
  };
}

export function removeChild(id) {
  return {
    type: REMOVE_CHILD,
    id
  };
}

export function moveChild(sourceID, targetID) {
  return {
    type: MOVE_CHILD,
    targetID,
    sourceID
  };
}

export function updateStyle(sourceID, newStyle) {
  return {
    type: UPDATE_STYLE,
    sourceID,
    newStyle
  };
}
