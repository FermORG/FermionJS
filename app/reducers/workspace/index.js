import { ADD_CHILD, REMOVE_CHILD, MOVE_CHILD, DELETE_CHILD, UPDATE_STYLE } from '../../actions/workspace';
import { ADD_STATE, ADD_PROPS, ADD_STYLES, ADD_EVENTS, DELETE_STATE, DELETE_PROPS, DELETE_STYLES, DELETE_EVENTS } from '../../actions/config';
import { SET_ACTIVE_COMPONENT } from '../../actions/FileSystemActions';
import { WORKSPACE_ID } from './../../constants';
import addComponent from './addComponent';
// import removeComponent from './removeComponent';
import moveChild from './moveChild';
import addStateValue from './addStateValue';
import addPropsValue from './addPropsValue';
import addStyleValue from './addStyleValue';
import addEvent from './addEvent';
import deleteStateValue from './deleteStateValue';
import deletePropsValue from './deletePropsValue';
import deleteStylesValue from './deleteStylesValue';
import deleteEvent from './deleteEvent';
import setActiveComponent from './setActiveComponent';
import updateStyle from './updateStyle';
// /// TEST DATA /////

const defaultWorkspace = {
  componentCounter: 2,
  activeComponent: '0',
  components: {
    workspace: {
      id: WORKSPACE_ID,
      children: [0, 1],
      events: {},
    },
  },
  state: {},
  methods: {},
};

defaultWorkspace.components[0] = {
  id: 0,
  name: 'BlackBox',
  children: [],
  parentID: WORKSPACE_ID,
  props: {
    style: {
      position: 'relative',
      height: '100px',
      width: '100px',
      display: 'inline-block',
      backgroundColor: 'black',
      overflow: 'auto',
    },
  },
  events: { test: 'event test =]' }
};

defaultWorkspace.components[1] = {
  id: 1,
  name: 'BlueBox',
  children: [],
  parentID: WORKSPACE_ID,
  props: {
    style: {
      position: 'relative',
      height: '100px',
      width: '100px',
      display: 'inline-block',
      backgroundColor: 'blue',
      overflow: 'auto',
    },
  },
  events: { test: 'event test =]' }
};

// Children is just a list of ids
// Order doesnt matter here, children could be something like -> [4, 2, 6, 1]
// defaultWorkspace.children[0] = 0;
// defaultWorkspace.children[1] = 1;

// ////////

export default function workspace(state = defaultWorkspace, action) {
  switch (action.type) {
    case ADD_CHILD:
      return addComponent(state, action);
// currently unused
    // case REMOVE_CHILD:
    //   return removeComponent(state, action);

    case MOVE_CHILD:
      return moveChild(state, action);

    case ADD_STATE:
      return addStateValue(state, action.aState);

    case ADD_PROPS:
      return addPropsValue(state, action.prop, action.component);

    case ADD_STYLES:
      return addStyleValue(state, action.style, action.component);

    case ADD_EVENTS:
      return addEvent(state, action.event, action.component);

    case DELETE_STATE:
      return deleteStateValue(state, action.propKey);

    case DELETE_PROPS:
      return deletePropsValue(state, action.prop, action.component);

    case DELETE_STYLES:
      return deleteStylesValue(state, action.style, action.component);

    case DELETE_EVENTS:
      return deleteEvent(state, action.event, action.component);

    case SET_ACTIVE_COMPONENT:
      return setActiveComponent(state, action.component);

    case UPDATE_STYLE:
      return updateStyle(state, action);
    default:
      return state;
  }
}
