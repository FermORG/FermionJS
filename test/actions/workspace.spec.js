import { WORKSPACE_ID } from '../../app/constants';
import { ADD_CHILD, CREATE_CHILD, DELETE_CHILD, MOVE_CHILD, REMOVE_CHILD, UPDATE_STYLE } from '../../app/actions/workspace';
import { addChild, removeChild, moveChild, updateStyle } from '../../app/actions/workspace';

describe('workspace actions', () => {
  describe('addChild', () => {
    it('should return a payload with the correct action type, a component and a target', () => {
      const componentMock = {
        name: 'testComponent',
        props: {
          testKey: 'testProps'
        },
      };
      const result = addChild('target', componentMock);
      expect(result).toMatchSnapshot();
      expect(result).toHaveProperty('targetID', 'target');
      expect(result).toHaveProperty('newComponent');
      expect(result).toHaveProperty('newComponent.props.testKey', 'testProps');
      expect(result).toHaveProperty('newComponent.name', 'testComponent');
      expect(result).toHaveProperty('newComponent.parentID', WORKSPACE_ID);
      expect(result).toHaveProperty('type', ADD_CHILD);
    });
  });

  describe('removeChild', () => {
    it('should return a payload with the correct action type and an id', () => {
      const result = removeChild('testID');
      expect(result).toMatchSnapshot();
      expect(result).toHaveProperty('type', REMOVE_CHILD);
      expect(result).toHaveProperty('id', 'testID');
    });
  });

  describe('moveChild', () => {
    it('should return a payload with the correct action type, a target and a source', () => {
      const result = moveChild('testSource', 'testTarget');
      expect(result).toMatchSnapshot();
      expect(result).toHaveProperty('type', MOVE_CHILD);
      expect(result).toHaveProperty('sourceID', 'testSource');
      expect(result).toHaveProperty('targetID', 'testTarget');
    });
  });

  describe('updateStyle', () => {
    it('should return a payload with correct action type, a source and a style', () => {
      const styleMock = {
        'height': '100px',
      };

      const result = updateStyle('testSource', styleMock);
      expect(result).toMatchSnapshot();
      expect(result).toHaveProperty('type', UPDATE_STYLE);
      expect(result).toHaveProperty('sourceID', 'testSource');
      expect(result).toHaveProperty('newStyle.height', '100px');
    });
  });

});
