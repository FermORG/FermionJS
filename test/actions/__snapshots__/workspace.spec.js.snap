// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workspace actions addChild should return a payload with the correct action type, a component and a target 1`] = `
Object {
  "newComponent": Object {
    "children": Array [],
    "events": undefined,
    "name": "testComponent",
    "parentID": "workspace",
    "props": Object {
      "testKey": "testProps",
    },
  },
  "targetID": "target",
  "type": "ADD_CHILD",
}
`;

exports[`workspace actions addComponent should return a payload with the correct action type, a component and a target 1`] = `
Object {
  "newComponent": Object {
    "children": Array [],
    "events": undefined,
    "name": "testComponent",
    "parentID": "workspace",
    "props": Object {
      "testKey": "testProps",
    },
  },
  "targetID": "target",
  "type": "ADD_COMPONENT",
}
`;

exports[`workspace actions deleteComponent should return a payload with the correct action type and an id 1`] = `
Object {
  "id": "testID",
  "type": "DELETE_COMPONENT",
}
`;

exports[`workspace actions moveChild should return a payload with the correct action type, a target and a source 1`] = `
Object {
  "sourceID": "testSource",
  "targetID": "testTarget",
  "type": "MOVE_CHILD",
}
`;

exports[`workspace actions moveComponent should return a payload with the correct action type, a target and a source 1`] = `
Object {
  "sourceID": "testSource",
  "targetID": "testTarget",
  "type": "MOVE_COMPONENT",
}
`;

exports[`workspace actions removeChild should return a payload with the correct action type and an id 1`] = `
Object {
  "id": "testID",
  "type": "REMOVE_CHILD",
}
`;

exports[`workspace actions updateStyle should return a payload with correct action type, a source and a style 1`] = `
Object {
  "newStyle": Object {
    "height": "100px",
  },
  "sourceID": "testSource",
  "type": "UPDATE_STYLE",
}
`;
